movingpandas/cpa.py

```diff
@@ -0,0 +1,412 @@
+        # This is the relative time (as a fraction of t1 - t0)
+        # when objects are closest together.
+        # In nautical applications is referred to as time to/until closest point
+        # of approach, e.g. http://dx.doi.org/10.12716/1001.09.01.06
+        t_tot = -np.dot(w0, dv).item() / dv2
+
+        # We introduce the concept of status so we can recall if objects are
+        # diverging or not.
+        status = ""
+
+        # Now that we know the fractional time to the closest point of approach we can
+        # determine if trajectories are moving closer together (converging) or
+        # away from each other (diverging), or in the time window t0 - t1, are
+        # approaching (reaching the closest point).
+        #
+        # Compute t as the clipped version of t_tot when converging or
+        # diverging (naming of t_tot, t are in line with naming in postgis / lwgeom)
+        #
+        t = t_tot
+        if t_tot > 1:
+            # converging
+            t = 1
+            status = "converging"
+        elif t < 0.0:
+            # diverging
+            t = 0
+            status = "diverging"
+        else:
+            status = "approaching"
+
+        # Create new coordinates when moving over the duration of
+        # We'll create a separate variable.
+        p_coords = p0 + pv * t
+        q_coords = q0 + qv * t
+
+        # p and q are the position at the cpa
+        p = shapely.Point(*p_coords)
+        q = shapely.Point(*q_coords)
+
+        # We can now also compute time of the closest point of approach (also
+        # referred to as TCPA in aerial applications) We refer to t_at
+        # (datetime) as time at the closest point of approach and t_to as the
+        # time to the closest point of approach. In python you can operate on
+        # datetimes with fractions (t1 - t0) is a time delta, t is a fraction
+        # Added to t0 this gives a datetime. This saves us from converting to
+        # timestamps and back. t = t0 + (t1 - t0) * t;
+
+        # t_to is not clipped to the current time window
+        t_to = (t1 - t0) * t_tot
+        # t_at is clipped, so we use t
+        t_at = t0 + (t1 - t0) * t
+        # make sure we have a timestamp
+        t_at = pd.Timestamp(t_at)
+
+        # Now we can compute the distance Because we do everything using coords
+        # this also works when geometry contains a z coordinate For example for
+        # airplanes and submarines.
+        d_pq = np.asarray(p.coords)[0] - np.asarray(q.coords)[0]
+        dist2 = np.dot(d_pq, d_pq).item()
+        dist = dist2**0.5
+
+        # Geopandas (in particular certain gis formats) do not support multiple
+        # geometry columns. So we'll return the closest point of approach as the
+        # closest line of approach. The line runs from a point on p0-p1 to a
+        # point on q0-q1.
+        pq = shapely.LineString([p, q])
+
+        # TODO: I think this should be multiplied with the duration of the
```

Comment:
Still working on this?

Reply:
This is how I implemented it. TODO removed.